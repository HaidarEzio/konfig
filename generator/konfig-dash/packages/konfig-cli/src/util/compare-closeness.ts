function calculateCloseness(target: string, string: string): number {
  const m = target.length
  const n = string.length

  // Check if string contains target as an exact word
  if (target.includes(string)) {
    // Give higher precedence to strings containing target as an exact word
    return -1
  }

  // Create a 2D array for the dynamic programming approach
  const dp: number[][] = Array.from({ length: m + 1 }, () =>
    Array(n + 1).fill(0)
  )

  // Initialize the first row and column of the array
  for (let i = 0; i <= m; i++) {
    dp[i][0] = i
  }

  for (let j = 0; j <= n; j++) {
    dp[0][j] = j
  }

  // Fill in the array using the Levenshtein distance algorithm
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (target[i - 1] === string[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1]
      } else {
        dp[i][j] =
          Math.min(
            dp[i - 1][j - 1], // substitution
            dp[i][j - 1], // insertion
            dp[i - 1][j] // deletion
          ) + 1
      }
    }
  }

  return dp[m][n]
}

export const compareCloseness = (
  target: string,
  stringA: string,
  stringB: string
): number => {
  const distanceA = calculateCloseness(target, stringA)
  const distanceB = calculateCloseness(target, stringB)

  if (distanceA < distanceB) {
    return -1
  } else if (distanceA > distanceB) {
    return 1
  } else {
    return 0
  }
}

// const targetString = "example";
// const stringArray = ["example1", "example2", "another", "test"];

// const sortedArray = stringArray.sort((stringA, stringB) =>
//   compareCloseness(targetString, stringA, stringB)
// );

// console.log(sortedArray);
