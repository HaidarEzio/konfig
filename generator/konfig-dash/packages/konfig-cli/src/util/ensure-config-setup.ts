import * as os from 'node:os'
import * as path from 'node:path'
import * as fs from 'node:fs'

const ARCHIVE_DIR_NAME = 'archive'

export const PROGRESS_PATH = path.join(
  getKonfigHomeDirectoryPath({}),
  'progress.json'
)

export type HomeDirParams = {
  mode?: 'dev' | 'prod' | 'test'
}

function homeDirName({ mode }: HomeDirParams) {
  if (mode === undefined || mode === 'prod') return '.konfig'
  if (mode === 'dev') return '.konfig-dev'
  if (mode === 'test') return '.konfig-test'
  throw Error(`Unexpected mode: ${mode}`)
}

export function getKonfigHomeDirectoryPath(params: HomeDirParams): string {
  return path.join(os.homedir(), homeDirName(params))
}

export function getArchiveDirectoryPath(params: HomeDirParams): string {
  return path.join(getKonfigHomeDirectoryPath(params), ARCHIVE_DIR_NAME)
}

export function ensureConfigSetup(params: HomeDirParams): void {
  if (!fs.existsSync(getKonfigHomeDirectoryPath(params)))
    fs.mkdirSync(getKonfigHomeDirectoryPath(params), { recursive: true })
}
