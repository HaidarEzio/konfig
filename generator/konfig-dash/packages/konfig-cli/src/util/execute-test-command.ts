import { CliUx } from '@oclif/core'
import execa, { ExecaSyncReturnValue, SyncOptions } from 'execa'
import { RequiredEnvironmentVariablesConfig, TestConfig } from 'konfig-lib'
import path from 'path'
import { parseKonfigYaml } from './parse-konfig-yaml'
import { parseFilterFlag } from './parseFilterFlag'

const validateRequiredEnvironmentVariables = (
  config: RequiredEnvironmentVariablesConfig
) => {
  if (config.requiredEnvironmentVariables === undefined) return
  for (const environmentVariable of config.requiredEnvironmentVariables) {
    if (!(environmentVariable in process.env))
      CliUx.ux.error(
        `Missing required environment variable: ${environmentVariable}`
      )
  }
}

export async function executeTestCommand({
  filterInput,
  sequence,
}: {
  filterInput?: string
  sequence: boolean
}) {
  const filter = parseFilterFlag(filterInput)
  const configDir = process.cwd()
  const { generators, common, additionalGenerators } = parseKonfigYaml({
    configDir,
  })
  validateRequiredEnvironmentVariables(common)
  const results: {
    generatorName: string
    result: ExecaSyncReturnValue<string>
  }[] = []
  const asyncResults: Promise<{
    generatorName: string
    result: ExecaSyncReturnValue<string>
  }>[] = []

  if (!sequence) CliUx.ux.log('Running tests in parallel')
  else CliUx.ux.log('Running tests sequentially')

  function runTest({
    test,
    outputDirectory,
    disabled,
    generatorName,
  }: {
    test?: TestConfig['test']
    outputDirectory?: string
    disabled?: boolean
    generatorName: string
  }) {
    if (filter !== null && !filter.includes(generatorName)) return
    if (disabled) return
    if (test === undefined) return
    if (outputDirectory === undefined)
      throw Error(
        `Generator "${generatorName}" must provide "outputDirectory" for "testScript" to work`
      )
    if (sequence)
      CliUx.ux.action.start(`Running tests for ${generatorName} SDK`)

    // verify environment variables are set if specified
    validateRequiredEnvironmentVariables(test)

    const commandOptions: SyncOptions = {
      cwd: path.join(configDir, outputDirectory),
      stdio: 'inherit',
      shell: true,
    }

    if (sequence) {
      // Sequential
      for (const idx in test.script) {
        const result = execa.commandSync(test.script[idx], commandOptions)
        if (parseInt(idx) == test.script.length - 1) {
          results.push({
            generatorName,
            result,
          })
        }
      }
    } else {
      // Parallel
      asyncResults.push(
        new Promise(async (resolve) => {
          for (const idx in test.script) {
            const result = await execa.command(test.script[idx], commandOptions)
            if (parseInt(idx) === test.script.length - 1) {
              resolve({ generatorName, result })
            }
          }
        })
      )
    }
    if (sequence) CliUx.ux.action.stop()
  }

  for (const [generatorName, generatorConfig] of [
    ...Object.entries(generators),
    ...(additionalGenerators ? Object.entries(additionalGenerators) : []),
  ]) {
    runTest({ ...generatorConfig, generatorName })
  }

  results.push(...(await Promise.all(asyncResults)))

  if (results.length === 0) {
    CliUx.ux.error('No tests were run')
  }

  let anyFailed: boolean = false
  for (const { result, generatorName } of results) {
    if (!result.failed) continue
    CliUx.ux.warn(`tests for ${generatorName} failed`)
  }
  if (anyFailed) CliUx.ux.error('Test failure occurred')

  // If we made it here then we successfully ran all tests
  CliUx.ux.info('Successfully ran all tests!')
}
