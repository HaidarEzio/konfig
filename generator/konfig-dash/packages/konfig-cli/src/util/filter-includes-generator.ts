import { KonfigYamlType } from 'konfig-lib'

export const filterIncludesGenerator = (
  filter: string[] | null,
  generator: keyof KonfigYamlType['generators']
) => {
  if (filter === null) return true
  return filter.includes(generator)
}
