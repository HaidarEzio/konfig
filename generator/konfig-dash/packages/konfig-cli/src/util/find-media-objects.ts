import {
  getOperations,
  MediaObject,
  ReferenceObject,
  resolveRef,
  Spec,
} from 'konfig-lib'

export function findMediaObjects({ spec }: { spec: Spec }) {
  const mediaObjectOrRefs: {
    media: MediaObject
    type: 'request' | 'response'
  }[] = []
  const operations = getOperations({ spec: spec.spec })

  // Request Body
  for (const { operation } of operations) {
    if (operation.requestBody === undefined) continue
    const requestBody = resolveRef({
      refOrObject: operation.requestBody,
      $ref: spec.$ref,
    })
    for (const media of Object.keys(requestBody.content)) {
      const mediaObjectOrRef = requestBody.content[media]
      mediaObjectOrRefs.push({ media: mediaObjectOrRef, type: 'request' })
    }
  }

  // Responses
  for (const { operation, path, method } of operations) {
    for (const statusCode of Object.keys(operation.responses)) {
      const responseObjectOrRef = operation.responses[statusCode]
      const responseObject = resolveRef({
        refOrObject: responseObjectOrRef,
        $ref: spec.$ref,
      })
      if (responseObject.content === undefined) continue
      for (const media of Object.keys(responseObject.content)) {
        const mediaObjectOrRef = responseObject.content[media]
        mediaObjectOrRefs.push({ media: mediaObjectOrRef, type: 'response' })
      }
    }
  }
  return mediaObjectOrRefs
}
