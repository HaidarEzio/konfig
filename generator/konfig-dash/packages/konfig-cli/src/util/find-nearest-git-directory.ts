import * as path from 'path'
import findUp from 'find-up'

export const findNearestGitDirectory = async ({
  cwd,
}: {
  cwd: string
}): Promise<string | undefined> => {
  return await findUp(
    (directory) => {
      const exists = findUp.sync.exists(path.join(directory, '.git'))
      return exists ? directory : undefined
    },
    { type: 'directory' }
  )
}
