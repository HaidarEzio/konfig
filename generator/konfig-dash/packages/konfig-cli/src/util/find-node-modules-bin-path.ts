import findUp from 'find-up'

export const findNodeModulesBinPath = async ({
  name,
  cwd,
}: {
  name: string
  cwd: string
}): Promise<string> => {
  const path = await findUp(`node_modules/.bin/${name}`, { cwd })
  if (path === undefined) throw Error(`Could not find path to '${name}'`)
  return path
}
