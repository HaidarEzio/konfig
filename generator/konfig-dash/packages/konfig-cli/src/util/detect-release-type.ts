import { CliUx } from '@oclif/core'
import { KonfigYamlType } from 'konfig-lib'
import { findNodeModulesBinPath } from './find-node-modules-bin-path'
import { getParentBranch } from './get-parent-branch'
import { simpleGit } from 'simple-git'
import * as shell from 'shelljs'
import semver from 'semver'

export async function detectReleaseType({
  cwd,
  specPath,
  baseBranchName,
}: {
  cwd: string
  specPath: string
  baseBranchName?: string
}): Promise<semver.ReleaseType> {
  const base = baseBranchName ? baseBranchName : getParentBranch()
  if (base === '')
    CliUx.ux.error(
      'Must be in Git repository and branched off of another branch to perform a diff.'
    )
  CliUx.ux.info(`Detected base branch: ${base}`)

  const isInGitHubAction = process.env['GITHUB_ACTIONS'] === 'true'
  if (isInGitHubAction) {
    CliUx.ux.info('Detected GitHub Action environment')
  }

  const opticPath = await findNodeModulesBinPath({
    name: 'optic',
    cwd,
  })

  if (specPath === undefined) CliUx.ux.error('Specify specPath konfig.yaml')
  CliUx.ux.info(
    `Picked up path to OpenAPI Spec at "./${specPath}" from konfig.yaml`
  )

  CliUx.ux.info(`Fetching and checking out "${base}" locally`)
  const git = simpleGit()
  await git.fetch()
  const currentBranch = isInGitHubAction
    ? (await git.log()).latest?.hash
    : await git.revparse(['--abbrev-ref', 'HEAD'])
  if (currentBranch === undefined)
    throw Error("Couldn't detect current branch name")
  CliUx.ux.info(`Detected current ref to be "${currentBranch}"`)
  CliUx.ux.action.start(`Checking out "${base}"`)
  await git.checkout(base)
  CliUx.ux.action.stop()
  CliUx.ux.action.start(`Checking out "${currentBranch}"`)
  await git.checkout(currentBranch)
  CliUx.ux.action.stop()

  // Detect breaking change
  CliUx.ux.action.start('Detecting potential breaking change')
  const command = `${opticPath} diff ${specPath} --check --base ${base}`
  const result = shell.exec(command)
  const releaseType: semver.ReleaseType = result.code === 1 ? 'major' : 'minor'
  CliUx.ux.action.stop()
  return releaseType
}
