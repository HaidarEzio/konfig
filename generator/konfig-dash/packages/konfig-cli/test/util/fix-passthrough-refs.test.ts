import { Spec, parseSpec, recurseObject } from 'konfig-lib'
import { fixPassthroughRefs } from '../../src/util/fix-passthrough-refs'

describe('fixPassthroughRefs', () => {
  it(`request body shouldn't be shortened`, async () => {
    const openapi: Spec['spec'] = {
      openapi: '3.0.0',
      info: {
        title: 'Test API',
        version: '1.0.0',
      },
      paths: {
        '/test': {
          post: {
            requestBody: {
              $ref: '#/components/requestBodies/RequestBody',
            },
            responses: {
              '200': {
                description: 'Successful response',
              },
            },
          },
        },
      },
      components: {
        requestBodies: {
          TestRequestBody: {
            content: {
              'application/json': {
                schema: {
                  $ref: '#/components/schemas/RequestBody',
                },
              },
            },
          },
        },
        schemas: {
          RequestBody: {
            $ref: '#/components/schemas/PassthroughSchema',
          },
          PassthroughSchema: {
            type: 'object',
            properties: {
              test: {
                type: 'string',
              },
            },
          },
        },
      },
    }

    const spec = await parseSpec(JSON.stringify(openapi))

    const fixedSpec = await fixPassthroughRefs({ spec })

    // snapshot expect
    expect(spec.spec).toMatchSnapshot()
  })
})
