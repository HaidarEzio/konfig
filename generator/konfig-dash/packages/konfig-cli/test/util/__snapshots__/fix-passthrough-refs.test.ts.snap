// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixPassthroughRefs request body shouldn't be shortened 1`] = `
{
  "components": {
    "requestBodies": {
      "TestRequestBody": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/PassthroughSchema",
            },
          },
        },
      },
    },
    "schemas": {
      "PassthroughSchema": {
        "properties": {
          "test": {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/test": {
      "post": {
        "requestBody": {
          "$ref": "#/components/requestBodies/RequestBody",
        },
        "responses": {
          "200": {
            "description": "Successful response",
          },
        },
      },
    },
  },
}
`;
